micro-rdk/src/esp32/log.rs

```diff
@@ -0,0 +1,129 @@
+//! This is for infrastructure supporting uploading logs from an ESP32 to Viam's cloud infrastructure
+//! (see common/log.rs for more details). On an ESP32, there are two sources of logs: logs from ESP-IDF,
+//! and logs from various Rust crates. The EspLogger from esp-idf-svc efficiently redirects the latter
+//! logs to UART in a format matching the first, so it is sufficient to simply wrap it. However
+//! to capture ESP-IDF logs, it is necessary to use the replace the default vprintf function to write to LOG_BUFFER
+//! (again, see common/log.rs) before invoking the previously existing vprintf function in order to write to
+//! UART. We store the previous vprintf function in PREVIOUS_LOGGER and use esp_log_set_vprintf for this purpose.
+use crate::{
+    common::log::ViamLogEntry,
+    google::protobuf::{value::Kind, Struct, Value},
+    proto::common::v1::LogEntry,
+};
+
+use esp_idf_svc::log::EspLogger;
+use esp_idf_svc::sys::{esp_log_set_vprintf, va_list, vprintf_like_t};
+use printf_compat::output::display;
+use ringbuf::Rb;
+use std::{collections::HashMap, sync::OnceLock};
+use std::{ffi::c_char, sync::Mutex};
+
+use crate::common::log::{get_log_buffer, ViamLogAdapter};
+
+fn previous_logger() -> &'static Mutex<vprintf_like_t> {
+    static PREVIOUS_LOGGER: OnceLock<Mutex<vprintf_like_t>> = OnceLock::new();
+    PREVIOUS_LOGGER.get_or_init(|| Mutex::new(None))
```

Comment:
isn't `OnceLock` just the thread-safe version of `OnceCell`?

Reply:
Yes I mistyped :P
Also do you need the Mutex here? the logging API is marked as thread safe since it uses the VFS api to write to STDOUT so we should be able to use the original v_printf without guarding it with a mutex?

Reply:
I probably don't need the mutex here, let me double-check